pkg/controllers/namespace/namespace_controller.go

```diff
@@ -225,32 +198,26 @@ func (r *NamespaceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 
 		ctx := context.Background()
 		annotations := obj.GetAnnotations()
-		err := r.insertNamespaceNetworkconfigBinding(ns, annotations)
+
+		ncName, err := r.VPCService.GetNetworkconfigNameFromAnnotation(ns, annotations)
```

Comment:
could we shelve the annotation related logics into the service method:
eg:

r.VPCService.GetNetworkConfigNameByNamespace(ns)

hide the get k8s ns obj, get ns obj annoation logic into service method.
in this way, maybe in controller logic, we only need to care reconcile part, and don't need to retrieve k8s objs.
I see in following lines for VPC service interface, you already impl it this way, hence we do not need to leave this code still in reconciler anymore.

Reply:
Sounds good to me to expose GetNetworkconfigNameFromNS instead of GetNetworkconfigNameFromAnnotation. Updated, thanks!
But we may still need to retrieve k8s objs in controller to determine whether this is a deleted ns.